From 11858441b7147142ca804bc9af17fd8bb0d06e54 Mon Sep 17 00:00:00 2001
From: Haris Okanovic <haris.okanovic@ni.com>
Date: Mon, 25 Feb 2019 17:31:50 -0600
Subject: [PATCH] Revert "global_init: ignore CURL_GLOBAL_SSL's absense"

This reverts commit d661b0afb571a68ec11f3dd856aba304c80780ba.

Restore CURL_GLOBAL_SSL flag in NI-cURL to maintain
backward compatibility.

Upstream-Status: Inappropriate [deprecated behavior]
---
 docs/libcurl/curl_global_init.3 |  4 ++--
 include/curl/curl.h             |  2 +-
 lib/easy.c                      | 20 ++++++++++++++------
 3 files changed, 17 insertions(+), 9 deletions(-)

diff --git a/docs/libcurl/curl_global_init.3 b/docs/libcurl/curl_global_init.3
index 6ba0920c1..ef4227fbb 100644
--- a/docs/libcurl/curl_global_init.3
+++ b/docs/libcurl/curl_global_init.3
@@ -62,8 +62,8 @@ Initialize everything possible. This sets all known bits except
 \fBCURL_GLOBAL_ACK_EINTR\fP.
 
 .IP CURL_GLOBAL_SSL
-(This flag's presence or absence serves no meaning since 7.57.0. The
-description below is for older libcurl versions.)
+(This flag's presence or absence serves no meaning since 7.57.0 in
+mainline cURL. It's still observed by NI-cURL.)
 
 Initialize SSL.
 
diff --git a/include/curl/curl.h b/include/curl/curl.h
index d1f857994..efa65576e 100644
--- a/include/curl/curl.h
+++ b/include/curl/curl.h
@@ -2569,7 +2569,7 @@ typedef enum {
   CURLCLOSEPOLICY_LAST /* last, never use this */
 } curl_closepolicy;
 
-#define CURL_GLOBAL_SSL (1<<0) /* no purpose since since 7.57.0 */
+#define CURL_GLOBAL_SSL (1<<0) /* Still observed by NI-cURL; no purpose since since 7.57.0 in mainline cURL */
 #define CURL_GLOBAL_WIN32 (1<<1)
 #define CURL_GLOBAL_ALL (CURL_GLOBAL_SSL|CURL_GLOBAL_WIN32)
 #define CURL_GLOBAL_NOTHING 0
diff --git a/lib/easy.c b/lib/easy.c
index ea267f809..f86e802d2 100644
--- a/lib/easy.c
+++ b/lib/easy.c
@@ -227,13 +227,16 @@ static CURLcode global_init(long flags, bool memoryfuncs)
 #endif
   }
 
-  if(!Curl_ssl_init()) {
-    DEBUGF(fprintf(stderr, "Error: Curl_ssl_init failed\n"));
+  /* XXX CURL_GLOBAL_SSL was removed from mainline cURL,
+   * keeping for backward compatibility */
+  if(flags & CURL_GLOBAL_SSL)
+    if(!Curl_ssl_init()) {
+      DEBUGF(fprintf(stderr, "Error: Curl_ssl_init failed\n"));
 #ifdef CURL_THREAD_SAFE_INIT
-    curl_global_mutex_unlock();
+      curl_global_mutex_unlock();
 #endif
-    return CURLE_FAILED_INIT;
-  }
+      return CURLE_FAILED_INIT;
+    }
 
   if(flags & CURL_GLOBAL_WIN32)
     if(win32_init()) {
@@ -369,7 +372,12 @@ void curl_global_cleanup(void)
   }
 
   Curl_global_host_cache_dtor();
-  Curl_ssl_cleanup();
+
+  /* XXX CURL_GLOBAL_SSL was removed from mainline cURL,
+   * keeping for backward compatibility */
+  if(init_flags & CURL_GLOBAL_SSL)
+    Curl_ssl_cleanup();
+
   Curl_resolver_global_cleanup();
 
   if(init_flags & CURL_GLOBAL_WIN32)
-- 
2.20.1

